File: frontend/src/types/contract.ts
export type ContractApprovalStatus = 'pending' | 'in_progress' | 'approved' | 'returned'

// 合同数据类型定义
export interface Contract {
  id: string
  teacher_code: string
  department: string
  name: string
  position: string
  gender: string
  age: number
  nation: string
  political_status: string
  id_number: string
  birthplace: string
  entry_date: string | null
  regular_date: string | null
  contract_start: string | null
  contract_end: string | null
  job_status: string
  approval_status: ContractApprovalStatus
  approval_completed_at: string | null
  resign_date: string | null
  phone_number: string
  education: string
  graduation_school: string
  diploma_no: string
  graduation_date: string | null
  major: string
  degree: string
  degree_no: string
  teacher_cert_type: string
  teacher_cert_no: string
  title_rank: string
  title_cert_no: string
  title_cert_date: string | null
  start_work_date: string | null
  teaching_years: number
  psychology_cert: string
  certificate_type: string
  teaching_grade: string
  teaching_subject: string
  last_work: string
  address: string
  emergency_contact: string
  emergency_phone: string
  mandarin_level: string
  remarks: string
  file_url: string
  ocr_confidence: number
  created_at: string
  updated_at: string
}

export interface ContractTimelineEvent {
  id: string
  type:
    | 'uploaded'
    | 'ocr_completed'
    | 'reviewed'
    | 'entry'
    | 'regularized'
    | 'renewal'
    | 'change'
    | 'termination'
    | 'note'
    | 'other'
  title: string
  description?: string
  operator?: string
  created_at: string
  extra_data?: Record<string, any>
}

export interface ContractAttachment {
  id: string
  name: string
  file_url: string
  file_type: string
  uploaded_at: string
  uploader?: string
}

export interface ContractLogItem {
  id: string
  action: string
  operator?: string
  created_at: string
  detail?: string
  changes?: Array<{
    field: string
    field_label: string
    before?: string | null
    after?: string | null
  }>
}

export interface ContractLifecycleDetail {
  contract: Contract
  timeline: ContractTimelineEvent[]
  attachments: ContractAttachment[]
  logs: ContractLogItem[]
  summary?: ContractLifecycleSummary
}

export interface ContractLifecycleSummary {
  nextAction?: {
    type: 'renewal' | 'review' | 'termination' | 'none'
    due_date?: string
    days_left?: number
    message?: string
    label?: string
  }
  pendingReviewFields?: string[]
  warnings?: string[]
}

// OCR 识别结果
export interface OcrResult {
  contract: Partial<Contract>
  confidence: Record<string, number>
  raw_text: string
}

// 字段置信度
export interface FieldConfidence {
  field: string
  value: string
  confidence: number
  isLowConfidence: boolean
}

// 查询参数
export interface ContractQuery {
  page?: number
  page_size?: number
  department?: string
  job_status?: string
  search?: string
  ids?: string[]
  approval_status?: ContractApprovalStatus | "all"
  expiring_within_days?: number
}

// 分页响应
export interface PaginatedResponse<T> {
  data: T[]
  total: number
  page: number
  page_size: number
  total_pages: number
}

export interface DashboardStats {
  totalTeachers: number
  activeContracts: number
  expiringSoon: number
  pendingReview: number
  averageConfidence: number
  breakdown: {
    totalTeachersBreakdown: {
      regular: number
      partTime: number
    }
    probationTeachers: number
    expiringWithin90Days: number
    averageConfidence: number
    pendingReview: number
  }
  sidebarSummary?: SidebarSummary
}

export interface SidebarSummary {
  pendingReview: number
  probationTeachers: number
  expiringWithin30Days: number
  expiringWithin90Days: number
  averageConfidence: number
}

export type ApprovalStatus = 'pending' | 'in_progress' | 'completed' | 'returned'

export type ApprovalPriority = 'high' | 'medium' | 'low'

export type ApprovalStageKey =
  | 'entry'
  | 'qualification'
  | 'probation'
  | 'signature'
  | 'renewal'
  | 'archive'

export interface ApprovalChecklistItem {
  label: string
  completed: boolean
}

export interface ApprovalTask {
  id: string
  contract_id: string
  teacher_name: string
  department: string
  stage: ApprovalStageKey
  status: ApprovalStatus
  priority: ApprovalPriority
  owner: string
  assignees: string[]
  due_date: string
  created_at: string
  updated_at: string
  check_items: ApprovalChecklistItem[]
  remarks?: string
  latest_action?: string
}

export interface ApprovalTaskQuery {
  status?: ApprovalStatus | 'all'
  stage?: ApprovalStageKey | 'all'
  keyword?: string
  page?: number
  page_size?: number
  filter_by_user?: boolean
}

export interface ApprovalTaskListResponse extends PaginatedResponse<ApprovalTask> {}

export interface ApprovalStatsOverview {
  pending: number
  in_progress: number
  completed: number
  returned: number
}

export interface ApprovalStageSummary {
  stage: ApprovalStageKey
  total: number
  pending: number
  completed: number
  overdue: number
}

export interface ApprovalHistoryRecord {
  id: string
  task_id: string
  action: string
  operator: string
  created_at: string
  comment?: string
}

export interface ApprovalActionPayload {
  comment?: string
}

export interface ApprovalActionResponse {
  task: ApprovalTask
}

File: backend/app/schemas/contract.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import date, datetime

# 基础 Schema
class ContractBase(BaseModel):
    teacher_code: str
    department: Optional[str] = None
    name: str
    position: Optional[str] = None
    gender: Optional[str] = None
    age: Optional[int] = None
    nation: Optional[str] = None
    political_status: Optional[str] = None
    id_number: Optional[str] = None
    birthplace: Optional[str] = None
    entry_date: Optional[date] = None
    regular_date: Optional[date] = None
    contract_start: Optional[date] = None
    contract_end: Optional[date] = None
    job_status: Optional[str] = '在职'
    resign_date: Optional[date] = None
    phone_number: Optional[str] = None
    education: Optional[str] = None
    graduation_school: Optional[str] = None
    diploma_no: Optional[str] = None
    graduation_date: Optional[date] = None
    major: Optional[str] = None
    degree: Optional[str] = None
    degree_no: Optional[str] = None
    teacher_cert_type: Optional[str] = None
    teacher_cert_no: Optional[str] = None
    title_rank: Optional[str] = None
    title_cert_no: Optional[str] = None
    title_cert_date: Optional[date] = None
    start_work_date: Optional[date] = None
    teaching_years: Optional[int] = 0
    psychology_cert: Optional[str] = None
    certificate_type: Optional[str] = None
    teaching_grade: Optional[str] = None
    teaching_subject: Optional[str] = None
    last_work: Optional[str] = None
    address: Optional[str] = None
    emergency_contact: Optional[str] = None
    emergency_phone: Optional[str] = None
    mandarin_level: Optional[str] = None
    remarks: Optional[str] = None
    file_url: Optional[str] = None
    ocr_confidence: Optional[float] = 0.0

# 创建 Schema
class ContractCreate(ContractBase):
    approval_status: Optional[str] = 'pending'

# 更新 Schema（所有字段可选）
class ContractUpdate(BaseModel):
    teacher_code: Optional[str] = None
    department: Optional[str] = None
    name: Optional[str] = None
    position: Optional[str] = None
    gender: Optional[str] = None
    age: Optional[int] = None
    nation: Optional[str] = None
    political_status: Optional[str] = None
    id_number: Optional[str] = None
    birthplace: Optional[str] = None
    entry_date: Optional[date] = None
    regular_date: Optional[date] = None
    contract_start: Optional[date] = None
    contract_end: Optional[date] = None
    job_status: Optional[str] = None
    resign_date: Optional[date] = None
    phone_number: Optional[str] = None
    education: Optional[str] = None
    graduation_school: Optional[str] = None
    diploma_no: Optional[str] = None
    graduation_date: Optional[date] = None
    major: Optional[str] = None
    degree: Optional[str] = None
    degree_no: Optional[str] = None
    teacher_cert_type: Optional[str] = None
    teacher_cert_no: Optional[str] = None
    title_rank: Optional[str] = None
    title_cert_no: Optional[str] = None
    title_cert_date: Optional[date] = None
    start_work_date: Optional[date] = None
    teaching_years: Optional[int] = None
    psychology_cert: Optional[str] = None
    certificate_type: Optional[str] = None
    teaching_grade: Optional[str] = None
    teaching_subject: Optional[str] = None
    last_work: Optional[str] = None
    address: Optional[str] = None
    emergency_contact: Optional[str] = None
    emergency_phone: Optional[str] = None
    mandarin_level: Optional[str] = None
    remarks: Optional[str] = None
    ocr_confidence: Optional[float] = None
    approval_status: Optional[str] = None

# 响应 Schema
class ContractResponse(ContractBase):
    id: str
    approval_status: str
    approval_completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ContractTimelineEvent(BaseModel):
    id: str
    event_type: str
    title: str
    description: Optional[str] = None
    operator: Optional[str] = None
    extra_data: Optional[dict[str, Any]] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ContractTimelineEventCreate(BaseModel):
    event_type: str
    title: str
    description: Optional[str] = None
    operator: Optional[str] = None
    extra_data: Optional[dict[str, Any]] = None


class ContractTimelineEventUpdate(BaseModel):
    event_type: Optional[str] = None
    title: Optional[str] = None
    description: Optional[str] = None
    extra_data: Optional[dict[str, Any]] = None


class ContractAttachmentItem(BaseModel):
    id: str
    name: str
    file_url: str
    file_type: str
    uploader: Optional[str] = None
    uploaded_at: datetime

    class Config:
        from_attributes = True


class ContractLogChangeItem(BaseModel):
    field: str
    field_label: str
    before: Optional[str] = None
    after: Optional[str] = None


class ContractLogItem(BaseModel):
    id: str
    action: str
    operator: Optional[str] = None
    detail: Optional[str] = None
    changes: Optional[List[ContractLogChangeItem]] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ContractLifecycleSummary(BaseModel):
    nextAction: Optional[dict[str, Any]] = None
    pendingReviewFields: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


class ContractLifecycleResponse(BaseModel):
    contract: ContractResponse
    timeline: List[ContractTimelineEvent]
    attachments: List[ContractAttachmentItem]
    logs: List[ContractLogItem]
    summary: Optional[ContractLifecycleSummary] = None


class ContractLifecycleRefreshResponse(BaseModel):
    timeline: List[ContractTimelineEvent]
    attachments: List[ContractAttachmentItem]
    logs: List[ContractLogItem]
    summary: Optional[ContractLifecycleSummary] = None

# 分页响应
class PaginatedResponse(BaseModel):
    data: list[Any]
    total: int
    page: int
    page_size: int
    total_pages: int


class ContractQuery(BaseModel):
    page: int = Field(1, ge=1)
    page_size: int = Field(20, ge=1, le=100)
    department: Optional[str] = None
    job_status: Optional[str] = None
    search: Optional[str] = None
    ids: Optional[List[str]] = None
    approval_status: Optional[str] = 'approved'

# OCR 结果
class OcrResult(BaseModel):
    contract: dict
    confidence: dict[str, float]
    raw_text: str
    low_confidence_fields: list[str] = Field(default_factory=list)
    original_filename: Optional[str] = None


class ContractTemplate(BaseModel):
    headers: list[str]
